include/pybind11/detail/internals.h

```diff
@@ -310,25 +310,38 @@ struct type_info {
 #    endif
 #endif
 
-/// On Linux/OSX, changes in __GXX_ABI_VERSION__ indicate ABI incompatibility.
-/// On MSVC, changes in _MSC_VER may indicate ABI incompatibility (#2898).
-#ifndef PYBIND11_BUILD_ABI
-#    if defined(__GXX_ABI_VERSION)
-#        define PYBIND11_BUILD_ABI "_cxxabi" PYBIND11_TOSTRING(__GXX_ABI_VERSION)
-#    elif defined(_MSC_VER)
-#        define PYBIND11_BUILD_ABI "_mscver" PYBIND11_TOSTRING(_MSC_VER)
+// Catch other conditions that imply ABI incompatibility
+// - MSVC builds with different CRT versions
+// - An anticipated MSVC ABI break ("vNext")
+// - Builds using libc++ with unstable ABIs
+// - Builds using libstdc++ with the legacy (pre-C++11) ABI
+#if defined(_MSC_VER)
+#    if defined(_MT) && defined(_DLL) // catches /MD or /MDd
+#        define PYBIND11_BUILD_LIB "_md"
+#    elif defined(_MT)
+#        define PYBIND11_BUILD_LIB "_mt" // catches /MT or /MTd
 #    else
+#        define PYBIND11_BUILD_LIB ""
+#    endif
+#    if (_MSC_VER) / 100 == 19
+#        define PYBIND11_BUILD_ABI NB_BUILD_LIB "_19"
+#    else
+#        define PYBIND11_BUILD_ABI NB_BUILD_LIB "_unknown"
+#    endif
+#elif defined(_LIBCPP_ABI_VERSION)
+#    define PYBIND11_BUILD_ABI "_abi" NB_TOSTRING(_LIBCPP_ABI_VERSION)
```

Comment:
`PYBIND11_TOSTRING`?

Reply:
Right :D. Should I update the PR, or will you adapt the changes in #5439?

I noticed that you changed some of the specific strings so that the generated ABI IDs are now incompatible with nanobind. (e.g. `_usecxx11` vs `_legacy` used here)

Reply:
Ah, I didn't realize.

Could you maybe work on #5439 directly? (Maybe after git merge master; sorry I forgot to update the PR today.)

My hope was that we can work together on 5439 until everybody is happy with it.

Reply:
I forgot to add:

I believe initially it'll be best to `#error` for unknown situations, so that we're sure we're properly covering everything that we're testing with today. As a last step we can add in "unknown" fallbacks.